kroxylicious-filters/kroxylicious-authorization/src/main/java/io/kroxylicious/filter/authorization/DeleteTopicsEnforcement.java

```diff
@@ -44,26 +51,59 @@ CompletionStage<RequestFilterResult> onRequest(RequestHeaderData header,
-                            throw authorizationFilter.topicIdsNotSupported();
-                        }
-                    }));
+            var partitioned = request.topics().stream().collect(Collectors.partitioningBy(t -> t.name() != null));
+            List<DeleteTopicsRequestData.DeleteTopicState> statesUsingIds = partitioned.get(false);
+            if (statesUsingIds.isEmpty()) {
+                actionsStage = CompletableFuture.completedStage(operation.actionsOf(partitioned.get(true).stream()
+                        .map(DeleteTopicsRequestData.DeleteTopicState::name)));
+                mappingStage = null;
+            }
+            else {
+                mappingStage = new CompletableFuture<>();
+                context.topicNames(statesUsingIds.stream()
+                        .map(DeleteTopicsRequestData.DeleteTopicState::topicId)
+                        .toList()).whenComplete((mapping, error) -> {
+                            // painfully, the stage returned by context.topicNames() cannot be converted to a future
+                            // so we have to make our own
+                            if (error == null) {
+                                mappingStage.complete(mapping);
+                            }
+                            else {
+                                mappingStage.completeExceptionally(error);
+                            }
+                        });
+                CompletionStage<List<Action>> actionsStage1 = mappingStage.thenApply(mapping -> {
+                    Stream<String> names = request.topics().stream()
+                            .map(t -> {
+                                return mapping.topicNames().get(t.topicId());
+                            });
+                    List<Action> actions = operation.actionsOf(names);
+                    return actions;
+                });
+                actionsStage = actionsStage1;
+            }
         }
         else {
-            actions = operation.actionsOf(request.topicNames().stream());
+            actionsStage = CompletableFuture.completedStage(operation.actionsOf(request.topicNames().stream()));
+            mappingStage = null;
         }
-        return authorizationFilter.authorization(context, actions)
+        return actionsStage
+                .thenCompose(actions -> authorizationFilter.authorization(context, actions))
                 .thenCompose(authorization -> {
                     if (useStates) {
-                        var decisions = authorization.partition(request.topics(), operation, DeleteTopicsRequestData.DeleteTopicState::name);
+                        Map<Decision, List<DeleteTopicsRequestData.DeleteTopicState>> decisions;
+                        if (mappingStage != null) {
+                            decisions = authorization.partition(request.topics(), operation,
+                                    state -> mappingStage.join().topicNames().get(state.topicId()));
```

Comment:
At this point we know the `mappingStage` must be done, because there's a chain over intermediate stages which have all completed.